source/slang/slang-capability.h

```diff
@@ -189,62 +131,172 @@ struct CapabilitySet
+            }
+            bool operator!=(const Iterator& other) const
+            {
+                return !(other == *this);
+            }
+
+            Iterator& operator++()
+            {
+                this->disjointSetNode++;
+                if (this->disjointSetNode == (*this->stageNode).second.disjointSets.end())
+                {
+nullDisjointSetNode:
+                    this->stageNode++;
+                    if (this->stageNode == (*this->targetNode).second.shaderStageSets.end())
+                    {
+nullStageSetNode:
+                        this->targetNode++;
+                        if (this->targetNode == this->context->end())
+                        {
+                            this->stageNode = {};
+                            this->disjointSetNode = {};
+                            return *this;
+                        }
+                        this->stageNode = (*this->targetNode).second.shaderStageSets.begin();
+                        if (this->stageNode == (*this->targetNode).second.shaderStageSets.end())
+                            goto nullStageSetNode;
+                    }
+
+                    this->disjointSetNode = (*this->stageNode).second.disjointSets.begin();
+                    if ((*this->stageNode).second.disjointSets.begin() == (*this->stageNode).second.disjointSets.end())
+                        goto nullDisjointSetNode;
+                }
+                return *this;
+            }
+            Iterator& operator++(int)
+            {
+                return ++(*this);
+            }
+            Iterator begin() const
+            {
+                // first node may not have a valid disjointSet or stageSet. handle the senario of: 
+                // 1. targetNode is empty
+                // 2. stage node has no valid stage sets (not empty)
+                // 3. stage node is empty
+                // 4. disjointSets has no sets
+                Iterator tmp(this->context);
+                tmp.targetNode = this->context->begin();
+                if (tmp.targetNode == this->context->end())
+                    return tmp;
+
+                tmp.stageNode = (*tmp.targetNode).second.shaderStageSets.begin();
+                while (tmp.stageNode == (*tmp.targetNode).second.shaderStageSets.end())
+                {
+trySetupStageNode:
+                    tmp.targetNode++;
+                    if (tmp.targetNode == this->context->end())
+                        return tmp;
+                    tmp.stageNode = (*tmp.targetNode).second.shaderStageSets.begin();
+                }
+                
+                tmp.disjointSetNode = (*tmp.stageNode).second.disjointSets.begin();
+                while (tmp.disjointSetNode == (*tmp.stageNode).second.disjointSets.end())
+                {
+                    tmp.stageNode++;
+                    if (tmp.stageNode == (*tmp.targetNode).second.shaderStageSets.end())
+                        goto trySetupStageNode;
```

Comment:
You can perhaps turn the `while` loop above into a `do-while` loop and use `continue` here. Again a lot of this can be simplified if we just get rid of the linked list.